### PR TITLE
Enable nullable reference types for IIncrementalAnalyzer

### PR DESCRIPTION
**Pull request by sharwell**

### CODE REVIEW

File: src/VisualStudio/Core/Def/Implementation/DesignerAttribute/DesignerAttributeIncrementalAnalyzer.cs [hunk 1]
```diff
@@ -61,15 +63,15 @@ internal partial class DesignerAttributeIncrementalAnalyzer : ForegroundThreadAf
         public Task DocumentResetAsync(Document document, CancellationToken cancellationToken)
         {
             _state.Remove(document.Id);
-            return _state.PersistAsync(document, new Data(VersionStamp.Default, VersionStamp.Default, designerAttributeArgument: null), cancellationToken);
+            return _state.PersistAsync(document, new Data(VersionStamp.Default, VersionStamp.Default, designerAttributeArgument: string.Empty), cancellationToken);
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Implementation/DesignerAttribute/DesignerAttributeIncrementalAnalyzer.cs]: 📝 Consumers of `Data` seem to assume that `DesignerAttributeArgument` is not null. It seemed easiest to just change this construction site. `CanReusePersistedVersion.CanReusePersistedVersion` will eventually prevent this value from being used here:

https://github.com/dotnet/roslyn/blob/1477e5781b4043d6e569becfdd2b6fd46c10dce2/src/VisualStudio/Core/Def/Implementation/DesignerAttribute/DesignerAttributeIncrementalAnalyzer.cs#L88-L96

File: src/Workspaces/Remote/Core/Telemetry/ApiUsageIncrementalAnalyzerProvider.cs [hunk 2]
```diff
@@ -76,6 +78,10 @@ public async Task AnalyzeProjectAsync(Project project, bool semanticsChanged, In
                 {
                     var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
                     var model = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                    if (model is null)
+                    {
+                        continue;
```
Comment (sharwell) [inline src/Workspaces/Remote/Core/Telemetry/ApiUsageIncrementalAnalyzerProvider.cs]: 📝 This is a new null check

File: src/Workspaces/Core/Portable/ExternalAccess/UnitTesting/UnitTestingIncrementalAnalyzer.cs [hunk 3]
```diff
@@ -15,8 +17,8 @@ internal class UnitTestingIncrementalAnalyzer : IIncrementalAnalyzer
         public UnitTestingIncrementalAnalyzer(IUnitTestingIncrementalAnalyzerImplementation implementation)
             => _implementation = implementation;
 
-        public Task AnalyzeDocumentAsync(Document document, SyntaxNode bodyOpt, InvocationReasons reasons, CancellationToken cancellationToken)
```
Comment (333fred) [inline src/Workspaces/Core/Portable/ExternalAccess/UnitTesting/UnitTestingIncrementalAnalyzer.cs]: Are these renames in the external access project not a breaking change?

Comment (sharwell) [reply src/Workspaces/Core/Portable/ExternalAccess/UnitTesting/UnitTestingIncrementalAnalyzer.cs]: Only the code in the [Api namespace](https://github.com/dotnet/roslyn/tree/a2d7737ebd3ed9896d1a2880d8c56335f9a6c7c2/src/Workspaces/Core/Portable/ExternalAccess/UnitTesting/Api) is externally visible.

**Comment by genlu** [comment]
>                 _solutionCrawlerService = workspace.Services.GetService<ISolutionCrawlerRegistrationService>() as SolutionCrawlerRegistrationService;

Could be 
`_registrationService as SolutionCrawlerRegistrationService`

---
Refers to: src/EditorFeatures/TestUtilities/Diagnostics/DiagnosticTaggerWrapper.cs:94 in a2d7737. [](commit_id = a2d7737ebd3ed9896d1a2880d8c56335f9a6c7c2, deletion_comment = False)

File: src/Features/Core/Portable/Diagnostics/DefaultDiagnosticAnalyzerService.cs [hunk 4]
```diff
@@ -229,9 +231,9 @@ public void RemoveProject(ProjectId projectId)
 
             private class DefaultUpdateArgsId : BuildToolId.Base<int, DocumentId>, ISupportLiveUpdate
             {
-                private readonly string _workspaceKind;
+                private readonly string? _workspaceKind;
 
-                public DefaultUpdateArgsId(string workspaceKind, AnalysisKind kind, DocumentId documentId) : base((int)kind, documentId)
+                public DefaultUpdateArgsId(string? workspaceKind, AnalysisKind kind, DocumentId documentId) : base((int)kind, documentId)
```
Comment (genlu) [inline src/Features/Core/Portable/Diagnostics/DefaultDiagnosticAnalyzerService.cs]: Noticed `AnalysisKind` is declared in both Feature and Workspace. Probably should move the one in Feature to `Microsoft.CodeAnalysis.Diagnostics.EngineV2` namespace?

File: src/Features/Core/Portable/Diagnostics/DiagnosticAnalyzerService_UpdateSource.cs [hunk 5]
```diff
@@ -19,13 +23,13 @@ internal partial class DiagnosticAnalyzerService : IDiagnosticUpdateSource
         private readonly EventMap _eventMap;
         private readonly SimpleTaskQueue _eventQueue;
 
-        private DiagnosticAnalyzerService(IDiagnosticUpdateSourceRegistrationService registrationService) : this()
+        private void InitializeDiagnosticAnalyzerService(IDiagnosticUpdateSourceRegistrationService registrationService, out EventMap eventMap, out SimpleTaskQueue eventQueue)
```
Comment (genlu) [inline src/Features/Core/Portable/Diagnostics/DiagnosticAnalyzerService_UpdateSource.cs]: static?

**Comment by genlu** [comment]
>                 var asyncToken = _owner?.Listener.BeginAsyncOperation(nameof(AnalyzeInProcAsync));

is `?` necessary?

---
Refers to: src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.InProcOrRemoteHostAnalyzerRunner.cs:83 in a2d7737. [](commit_id = a2d7737ebd3ed9896d1a2880d8c56335f9a6c7c2, deletion_comment = False)

File: src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.cs [hunk 6]
```diff
@@ -48,7 +48,7 @@ internal partial class DiagnosticIncrementalAnalyzer : IIncrementalAnalyzer
             int correlationId,
             Workspace workspace,
             DiagnosticAnalyzerInfoCache analyzerInfoCache,
-            AbstractHostDiagnosticUpdateSource hostDiagnosticUpdateSource)
+            AbstractHostDiagnosticUpdateSource? hostDiagnosticUpdateSource)
         {
             Contract.ThrowIfNull(analyzerService);
```
Comment (genlu) [inline src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.cs]: is this necessary?

File: src/Features/Core/Portable/IncrementalCaches/SymbolTreeInfoIncrementalAnalyzerProvider.cs [hunk 7]
```diff
@@ -231,14 +233,15 @@ private async Task UpdateSourceSymbolTreeInfoAsync(Project project, Cancellation
             }
 
             [PerformanceSensitive("https://github.com/dotnet/roslyn/issues/36158", AllowCaptures = false, Constraint = "Avoid captures to reduce GC pressure when running in the host workspace.")]
-            private Task GetTask(Project project, Func<IncrementalAnalyzer, Project, PortableExecutableReference, CancellationToken, Task> func, PortableExecutableReference reference, CancellationToken cancellationToken)
+            private Task GetTask<TReference>(Project project, Func<IncrementalAnalyzer, Project, TReference, CancellationToken, Task> func, TReference reference, CancellationToken cancellationToken)
+                where TReference : MetadataReference?
```
Comment (genlu) [inline src/Features/Core/Portable/IncrementalCaches/SymbolTreeInfoIncrementalAnalyzerProvider.cs]: why `PortableExecutableReference` is changed to `MetadataReference`?

File: src/Workspaces/Remote/Core/Telemetry/ApiUsageIncrementalAnalyzerProvider.cs [hunk 8]
```diff
@@ -154,7 +160,7 @@ public async Task AnalyzeProjectAsync(Project project, bool semanticsChanged, In
                     }
 
                     if (symbol.Locations.All(l => l.Kind == LocationKind.MetadataFile) &&
-                        solutionOpt?.GetProject(symbol.ContainingAssembly, cancellationToken) == null)
+                        solution?.GetProject(symbol.ContainingAssembly, cancellationToken) == null)
```
Comment (genlu) [inline src/Workspaces/Remote/Core/Telemetry/ApiUsageIncrementalAnalyzerProvider.cs]: I'm curious about the purpose of this analyzer. How is the collected data used?

File: src/Features/Core/Portable/IncrementalCaches/SymbolTreeInfoIncrementalAnalyzerProvider.cs (hunk 7 above)
Comment (sharwell) [reply src/Features/Core/Portable/IncrementalCaches/SymbolTreeInfoIncrementalAnalyzerProvider.cs]: The argument needs to get passed to the delegate with the same nullability (i.e. allow null for the argument, but if a non-null argument is passed then the delegate also sees a non-null argument). `PortableExecutableReference` is derived from `MetadataReference`, so the constraint was simplified.

File: src/Workspaces/Remote/Core/Telemetry/ApiUsageIncrementalAnalyzerProvider.cs (hunk 8 above)
Comment (sharwell) [reply src/Workspaces/Remote/Core/Telemetry/ApiUsageIncrementalAnalyzerProvider.cs]: ➡️ Discussion moved to #41114

File: src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.cs (hunk 6 above)
Comment (sharwell) [reply src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.cs]: Maybe we remove them after we finish enabling nullable reference types, but for now it doesn't seem to hurt.

File: src/Features/Core/Portable/Diagnostics/DefaultDiagnosticAnalyzerService.cs (hunk 4 above)
Comment (sharwell) [reply src/Features/Core/Portable/Diagnostics/DefaultDiagnosticAnalyzerService.cs]: Seems out of scope for this change.

**Comment by sharwell** [comment]
@genlu @dotnet/roslyn-ide for review

**Comment by sharwell** [comment]
Closing as out-of-date.

**Pull request closed without merging**